src/test/java/com/redhat/labs/lodestar/service/ProjectStructureServiceTest.java

```diff
@@ -0,0 +1,160 @@
+
+    @Inject
+    ProjectStructureService psService;
+
+    @InjectMock
+    @RestClient
+    GitLabService gitLabService;
+
+    // New Customers and Projects
+
+//    @Test
+//    void testThis() {
+//
+//        String customerName = "customer1";
+//        String projectName = "project1";
+//
+//        Group cGroup = Group.builder().name(customerName).path(GitLabPathUtils.generateValidPath(customerName))
+//                .parentId(null).build();
+//
+//        Group pGroup = Group.builder().name(projectName).path(GitLabPathUtils.generateValidPath(projectName))
+//                .parentId(null).build();
+//
+////        when(groupService.createGitLabGroup(ArgumentMatchers.argThat((g) -> g.getName().equals(customerName)))).thenReturn(Optional.of(cGroup));
+////        when(groupService.createGitLabGroup(ArgumentMatchers.argThat((g) -> g.getName().equals(customerName)))).thenReturn(Optional.of(pGroup));
+//        when(groupService.createGitLabGroup(Mockito.any(Group.class)))
+//                .thenAnswer((i) -> Optional.of(i.getArgument(0, Group.class)));
+//
+//        Optional<Group> group = psService.testThis(customerName);
+//        System.out.print(group.orElse(null));
+//
+//    }
+
+    @Test
+    void testCreateProjectStructureForNewCustomerAndNewProject() {
+
+        // given
+        String customerName = "customer1";
+        String projectName = "project1";
+
+        Group cGroup = Group.builder().name(customerName).path(GitLabPathUtils.generateValidPath(customerName)).id(2222)
+                .parentId(1111).build();
+        Group pGroup = Group.builder().name(projectName).path(GitLabPathUtils.generateValidPath(projectName))
+                .parentId(2222).id(3333).build();
+        Project project = Project.builder().id(4444).name("iac").visibility("private")
+                .namespace(Namespace.builder().id(3333).parentId(2222).build()).build();
+
+        given(gitLabService.getProjectById(Mockito.anyString())).willReturn(null);
+        given(gitLabService.getSubGroups(Mockito.anyInt(), Mockito.any(Integer.class), Mockito.any(Integer.class)))
+                .willReturn(Response.ok(Arrays.asList()).header("X-Total-Pages", 1).build());
+        given(gitLabService.createGroup(Mockito.any(Group.class))).willReturn(cGroup, pGroup);
+        given(gitLabService.createProject(Mockito.any(Project.class))).willReturn(project);
+
+        Engagement engagement = Engagement.builder().customerName(customerName).projectName(projectName).build();
+
+        // when
+        Project actual = psService.createOrUpdateProjectStructure(engagement, "http://some-path/engagements");
+
+        // then
+        assertNotNull(project);
+        assertEquals("iac", actual.getName());
+        assertEquals("private", actual.getVisibility());
+        assertEquals(3333, actual.getNamespace().getId());
+        assertEquals(2222, actual.getNamespace().getParentId());
+
+    }
+
+}
+
+//New Customers and Projects - 2/2
```

Comment:
same here. is this a todo?

Reply:
I was using this giant list as the scenarios I was manually testing.  Ideally these should have been implemented as automated tests, but looks like I forgot about it.  Those will have to be done later, when I have more time.

Reply:
ok. makes sense. It's a little tricky validating back to git in the e2e but I think it can be done. Maybe capture this in a doc outside the code and then somebody else can pick it up.